File: org/jackhuang/compactwatermills/InventorySlotConsumable.java
package org.jackhuang.compactwatermills;

import java.util.Random;

import org.jackhuang.compactwatermills.util.StackUtil;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;

public abstract class InventorySlotConsumable extends InventorySlot {
	private static final Random random = new Random();

	public InventorySlotConsumable(TileEntityInventory base, String name,
			int oldStartIndex, int count) {
		super(base, name, oldStartIndex, InventorySlot.Access.I, count,
				InventorySlot.InvSide.TOP);
	}

	public InventorySlotConsumable(TileEntityInventory base, String name,
			int oldStartIndex, InventorySlot.Access access, int count,
			InventorySlot.InvSide preferredSide) {
		super(base, name, oldStartIndex, access, count, preferredSide);
	}

	public abstract boolean accepts(ItemStack paramItemStack);

	public boolean canOutput() {
		return (super.canOutput())
				|| ((this.access != InventorySlot.Access.NONE)
						&& (get() != null) && (!accepts(get())));
	}

	public ItemStack consume(int amount) {
		return consume(amount, false, false);
	}

	public ItemStack consume(int amount, boolean simulate,
			boolean consumeContainers) {
		ItemStack ret = null;

		for (int i = 0; i < size(); i++) {
			ItemStack itemStack = get(i);

			if ((itemStack == null)
					|| (itemStack.stackSize < 1)
					|| (!accepts(itemStack))
					|| ((ret != null) && (!StackUtil.isStackEqual(itemStack,
							ret)))
					|| ((itemStack.stackSize != 1) && (!consumeContainers) && (itemStack
							.getItem().hasContainerItem()))) {
				continue;
			}

			int currentAmount = Math.min(amount, itemStack.stackSize);

			amount -= currentAmount;

			if (!simulate) {
				if (itemStack.stackSize == currentAmount) {
					if ((!consumeContainers)
							&& (itemStack.getItem().hasContainerItem()))
						put(i,
								itemStack.getItem().getContainerItemStack(
										itemStack));
					else
						put(i, null);
				} else {
					itemStack.stackSize -= currentAmount;
				}
			}

			if (ret == null)
				ret = StackUtil.copyWithSize(itemStack, currentAmount);
			else {
				ret.stackSize += currentAmount;
			}

			if (amount == 0) {
				break;
			}
		}
		return ret;
	}

	public ItemStack damage(int amount) {
		return damage(amount, null);
	}

	public ItemStack damage(int amount, EntityLivingBase src) {
		ItemStack ret = null;
		int damageApplied = 0;

		for (int i = 0; i < size(); i++) {
			ItemStack itemStack = get(i);

			if ((itemStack == null)
					|| (!accepts(itemStack))
					|| (!itemStack.getItem().isDamageable())
					|| ((ret != null) && ((itemStack.itemID != ret.itemID) || (!ItemStack
							.areItemStackTagsEqual(itemStack, ret))))) {
				continue;
			}
			int currentAmount = Math.min(amount, itemStack.getMaxDamage()
					- itemStack.getItemDamage());

			damageApplied += currentAmount;
			amount -= currentAmount;

			if (src != null)
				itemStack.damageItem(currentAmount, src);
			else {
				itemStack.attemptDamageItem(currentAmount, random);
			}

			if (itemStack.getItemDamage() >= itemStack.getMaxDamage()) {
				itemStack.stackSize -= 1;
				itemStack.setItemDamage(0);
			}

			if (itemStack.stackSize == 0)
				put(i, null);
			else {
				i--;
			}

			if (ret == null)
				ret = itemStack.copy();

			if (amount == 0) {
				break;
			}
		}
		ret.stackSize = (damageApplied / ret.getMaxDamage());
		ret.setItemDamage(damageApplied % ret.getMaxDamage());

		return ret;
	}
}

File: src/main/kotlin/waterpower/common/block/inventory/InventorySlotConsumable.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.inventory

import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import waterpower.common.block.tile.TileEntityInventory
import waterpower.util.*
import java.util.*

abstract class InventorySlotConsumable : InventorySlot {

    constructor(base: TileEntityInventory, name: String, count: Int) : super(base, name, InventorySlot.Access.I, count, InventorySlot.InvSide.TOP) {}

    constructor(base: TileEntityInventory, name: String, access: InventorySlot.Access, count: Int, preferredSide: InventorySlot.InvSide) : super(base, name, access, count, preferredSide) {}

    abstract override fun accepts(stack: ItemStack?): Boolean

    override fun canOutput(): Boolean {
        return super.canOutput() || this.access !== InventorySlot.Access.NONE && !isStackEmpty(get()) && !accepts(get()!!)
    }

    fun consume(count: Int, simulate: Boolean = false, consumeContainers: Boolean = false): ItemStack? {
        var amount = count
        var ret = emptyStack

        for (i in 0..size() - 1) {
            var itemStack = get(i)

            if (isStackEmpty(itemStack) || !accepts(itemStack!!) || !isStackEmpty(ret) && !isStackEqual(itemStack, ret)
                    || getCount(itemStack) != 1 && !consumeContainers && itemStack.item.hasContainerItem()) {
                continue
            }

            val currentAmount = minOf(amount, getCount(itemStack))

            amount -= currentAmount

            if (!simulate)
                if (getCount(itemStack) == currentAmount)
                    if (!consumeContainers && itemStack.item.hasContainerItem())
                        put(i, itemStack.item.getContainerItem(itemStack))
                    else
                        clear(i)
                else
                    itemStack = shrink(itemStack)

            if (isStackEmpty(ret))
                ret = itemStack!!.copyWithNewCount(currentAmount)
            else
                ret = grow(ret!!)

            if (amount == 0)
                break
        }
        return ret
    }

    fun damage(count: Int, src: EntityLivingBase? = null): ItemStack {
        var amount = count
        var ret = emptyStack
        var damageApplied = 0

        var i = 0
        while (i < size()) {
            var itemStack = get(i)

            if (isStackEmpty(itemStack) || !accepts(itemStack!!) || !itemStack.item.isDamageable
                    || !isStackEmpty(ret) && (itemStack.item !== ret!!.item || !ItemStack.areItemStackTagsEqual(itemStack, ret))) {
                ++i
                continue
            }
            val currentAmount = minOf(amount, itemStack.maxDamage - itemStack.itemDamage)

            damageApplied += currentAmount
            amount -= currentAmount

            if (src != null)
                itemStack.damageItem(currentAmount, src)
            else {
                itemStack.attemptDamageItem(currentAmount, random)
            }

            if (itemStack.itemDamage >= itemStack.maxDamage) {
                itemStack = shrink(itemStack)
                itemStack!!.itemDamage = 0
            }

            if (isStackEmpty(itemStack))
                put(i, emptyStack)
            else {
                i--
            }

            if (isStackEmpty(ret))
                ret = itemStack.copy()

            if (amount == 0) {
                break
            }
            ++i
        }
        ret!!.set(damageApplied / ret.maxDamage)
        ret.itemDamage = damageApplied % ret.maxDamage

        return ret
    }

    companion object {
        private val random = Random()
    }
}